trace/trace.go

```diff
@@ -215,21 +225,33 @@ func (s *Span) End() {
 	if !s.IsRecordingEvents() {
 		return
 	}
+
+	// If there are no exporters, no need to create spanData
+	// nor invoke all the code below this check, just return.
+	// See Issue 681.
+	if atomic.LoadInt64(&nExporters) <= 0 {
+		return
+	}
+
 	s.exportOnce.Do(func() {
-		// TODO: optimize to avoid this call if sd won't be used.
+		if !s.spanContext.IsSampled() {
```

Comment:
This breaks zpages (as well as the prior early return outside of the Do). You need to ensure that s.spanStore.finish gets called.

Reply:
Oh, I see, thanks for pointing that out.